extensions/amp-base-carousel/1.0/scroller.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as Preact from '../../../src/preact';
+import * as styles from './base-carousel.css';
+import {WithAmpContext} from '../../../src/preact/context';
+import {useEffect, useLayoutEffect, useRef} from '../../../src/preact';
+
+/**
+ * @param {!JsonObject} props
+ * @return {PreactDef.Renderable}
+ */
+export function Scroller(props) {
+  const {children, onSlideChange, currentSlide, setCurrentSlide} = props;
+  const slides = children || [];
+  const containerRef = useRef();
+
+  // Note: this has to be useLayoutEffect, not useEffect.
+  useLayoutEffect(() => {
+    if (!containerRef.current) {
+      return;
+    }
+    const container = containerRef.current;
+    container /* OK */.scrollLeft =
+      container /* OK */.offsetWidth * currentSlide;
+  }, [currentSlide]);
+
+  // Event. Don't need to be render-blocking.
+  useEffect(() => {
+    if (onSlideChange) {
+      onSlideChange({currentSlide});
+    }
+  }, [currentSlide, onSlideChange]);
+
+  /**
+   *
+   */
+  function scrollHandler() {
```

Comment:
`useCallback` is needed here.

Reply:
PTAL at recent changes and let me know if `useCallback` may still be needed here. I've modified this function to call a few methods which, if each of these utilized `useCallback`, would include `restingIndex` in a root dependency array. As it stands, `restingIndex` is what changes to trigger each `render` -- so in the common case this will effectively be defined on every render with or without `useCallback`.